keras_nlp/models/gpt2/gpt2_causal_lm.py

```diff
@@ -200,17 +213,138 @@ def backbone_cls(cls):
+            cache: a dense float Tensor, the cache of key and value.
+            cache_index: int, or int Tensor, defaults to None. If set, it
+                represents the index of current inputs in the whole sequence.
+
+        Returns:
+            x: a dense float Tensor, the next token logits of `inputs`.
+            cache: a dense float Tensor, the updated cache.
+        """
+        token_ids = inputs["token_ids"]
+        padding_mask = inputs["padding_mask"]
+        token_embedding = self.token_embeddings(token_ids)
+        if cache_index is None:
+            position_embedding = self.position_embeddings(token_embedding)
+        else:
+            position_embedding = self.position_embeddings.position_embeddings[
+                cache_index, :
+            ]
+        x = token_embedding + position_embedding
+        x = self.embeddings_dropout(x)
+        # Each `TransformerDecoder` layer has a cache, we update them
+        # separately.
+        caches = tf.unstack(cache, axis=1)
+        for i, transformer_layer in enumerate(self.transformer_layers):
+            current_cache = caches[i]
+            x, current_cache = transformer_layer(
+                x,
+                decoder_padding_mask=padding_mask,
+                cache=current_cache,
+                cache_index=cache_index,
+            )
+            caches[i] = current_cache
+        cache = tf.stack(caches, axis=1)
+        x = self.layer_norm(x)
+        x = tf.matmul(
+            x,
+            self.backbone.token_embedding.embeddings,
+            transpose_b=True,
+        )
+        return x, cache
+
+    def build_initial_cache(self, initial_inputs, max_length):
+        """Build initial cache based on the prompt.
+
+        This method should be called before the decoding loop to build the
+        initial cache. The cache is of shape [2, `self.num_layers`, batch_size,
+        max_length, `self.num_heads`, `self.hidden_dim // self.num_heads`].
+        The first dim represents it's a key or value in multi-head attention.
+
+        Args:
+            initial_inputs: a dense Tensor, the initial inputs to the decoding
+                loop.
+            max_length: int, the max length of the generated sequence.
+
+        Returns:
+            cache: a dense float Tensor, the cache of key and value.
+            max_length: int, the max length of generated sequence.
+        """
+        token_ids = initial_inputs["token_ids"]
+        padding_mask = initial_inputs["padding_mask"]
+
+        if max_length < self.backbone.max_sequence_length:
+            token_ids = token_ids[:, :max_length]
+            padding_mask = padding_mask[:, :max_length]
+
+        x = {
+            "token_ids": token_ids,
+            "padding_mask": padding_mask,
+        }
+
+        batch_size = tf.shape(token_ids)[0]
+        outputs = tf.zeros(
+            [batch_size, max_length, self.backbone.vocabulary_size]
+        )
+        cache = tf.zeros(
```

Comment:
Another thing I just remembered, we should update this cache shape so that `batch_size` comes first (will require updates elsewhere, e.g. stack/unstack axes).

The issue is that these axis choices are getting reflected all the way down to the cache layer, and even though we are not doing it today, we should support someone else writing a functional cache model. For functional models, the batch axis must come first.

Reply:
sg, let me change it.